Source/Core/VideoCommon/BPFunctions.h

```diff
@@ -7,16 +7,132 @@
 
 #pragma once
 
-#include "Common/MathUtil.h"
+#include <set>
+#include <utility>
 
-struct BPCmd;
+#include "Common/MathUtil.h"
+#include "VideoCommon/BPMemory.h"
+struct XFMemory;
 
 namespace BPFunctions
 {
+struct ScissorRange
+{
+  constexpr ScissorRange(u32 offset, u32 start, u32 end) : offset(offset), start(start), end(end) {}
+  const int offset;
+  const int start;
+  const int end;
+};
+
+struct ScissorRect
+{
+  constexpr ScissorRect(ScissorRange x_range, ScissorRange y_range)
+      :  // Rectangle ctor takes x0, y0, x1, y1.
+        rect(x_range.start, y_range.start, x_range.end, y_range.end), x_off(x_range.offset),
+        y_off(y_range.offset)
+  {
+  }
+
+  MathUtil::Rectangle<int> rect;
+  int x_off;
+  int y_off;
+
+  int GetArea() const;
+};
+
+struct ScissorResult
```

Comment:
I haven't seen any code documentation that explains why there are multiple scissor rects and how one is better than others among them. Maybe a docstring for this struct would be a good place to have this? (Discard if there is documentation already and I just happened to miss it)

Reply:
I've added some documentation here; let me know how it looks.

Reply:
Looks great, thanks!